src/libnpf/npf.c

```diff
@@ -1061,6 +1061,34 @@ npf_table_insert(nl_config_t *ncf, nl_table_t *tl)
 	return 0;
 }
 
+int
+npf_table_replace(nl_table_t *tl, int fd, npf_error_t *errinfo)
+{
+	nvlist_t *errnv = NULL;
+	int error;
+
+	/* Ensure the table is built. */
+	(void)_npf_table_build(tl);
```

Comment:
Table building is only applicable for the `NPF_TABLE_CONST` type of tables (they use perfect hashing). You can move the test (see the `npf_table_insert()` logic) to the `_npf_table_build()` function itself.

Reply:
I'm not totally sure what you mean by "moving the test" - which test do you mean?

I had envisaged that this implementation of table replacements could be used for `NPF_TABLE_CONST` tables also - it should still be able to take advantaged of perfect hashing since  it's building a  completely new table, rather than adding/removing entries to the existing table.  Or am I missing something here?

Reply:
The `_npf_table_build()` function builds a CDB regardless of the actual table type. This step is applicable if the table type is `NPF_TABLE_CONST`, meaning you need the following check (like in `npf_table_insert()`):
```c
if (dnvlist_get_number(tl->table_dict, "type", 0) == NPF_TABLE_CONST) {
	if ((error = _npf_table_build(tl)) != 0) {
		return error;
	}
}
```
Alternatively, you can move this check into the `_npf_table_build()` itself.

Reply:
Great, thanks for the pointer; I've renamed `_npf_table_build()` to `_npf_table_build_const()` and moved the `NPF_TABLE_CONST` test there rather than duplicate the logic.
Let me know if you want any changes.